### PR TITLE
[DEV-7584] Federal Account Quarterly Aggregation Endpoint Updates

### PR DESCRIPTION
**Pull request by collinwr**
**Description:**
This PR updates the Quarterly TAS Aggregation endpoints to take only the latest period of each quarter into account. Previously all Obligation/Outlay values from any period of each quarter were being included in the aggregation.

**Technical details:**
This is accomplished in a similar manner as the solution for a previous TAS endpoint bug: https://github.com/fedspendingtransparency/usaspending-api/pull/3191/files

A new helper method was added to return a list of submission ids from the latest period of each quarter for each Reporting Agency. This list of `submission_id`s used to narrow down the records used in the final aggregation. Two optional parameters are included `fiscal_years` and `federal_account_id`, to further narrow down the list of `submission_id`s. 

Note: The updated APIs are V1 APIs, and don't follow the same conventions of the newer, V2 APIs. This means that Django query strings are passed directly as a part of the API. The changes from this PR to the API include hardcoded checks for two of these query strings to pull specific filters. 

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. [x] API documentation updated
3. [ ] Necessary PR reviewers:
    - [ ] Backend
    - [ ] Frontend
4. [x] Matview impact assessment completed
5. [x] Frontend impact assessment completed
6. [x] Data validation completed
7. [ ] Appropriate Operations ticket(s) created
8. [x] Jira Ticket [DEV-7584](https://federal-spending-transparency.atlassian.net/browse/DEV-7584):
    - [x] Link to this Pull-Request
    - [ ] Performance evaluation of affected (API | Script | Download)
    - [ ] Before / After data comparison

**Area for explaining above N/A when needed:**
```
```

### CODE REVIEW

File: usaspending_api/common/mixins.py [hunk 1]
```diff
@@ -212,6 +212,30 @@ def order_records(self, request, *args, **kwargs):
         else:
             return queryset
 
+    def get_federal_account_id(self):
+        """
+        Returns the federal_account_id from the list of query parameters. If parameter
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/common/mixins.py]: This has been updated

File: usaspending_api/common/mixins.py [hunk 2]
```diff
@@ -212,6 +212,30 @@ def order_records(self, request, *args, **kwargs):
...
+        federal_account_id = None
+        if "filters" in self.request.data:
+            for filter in self.request.data["filters"]:
+                if filter["field"] == "treasury_account__federal_account_id":
+                    federal_account_id = filter["value"]
+        return federal_account_id
+
+    def get_fiscal_years(self):
+        """
+        Returns the list of fiscal_years from the list of query parameters. If parameter
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/common/mixins.py]: This has been updated

File: usaspending_api/submissions/helpers.py [hunk 3]
```diff
@@ -198,3 +198,36 @@ def get_latest_submission_ids_for_fiscal_year(fiscal_year: int):
...
+    Returns a list of submission_ids that consists of the latest submission_id for each quarter of a given
+    fiscal year and federal account. This list will capture cases where a Reporting Agency might not submit
+    in the most recent Submission Period but they do have a Submission in the provided Fiscal Year.
+    """
+
+    filters = {"submission_window__submission_reveal_date__lte": now()}
+    if len(fiscal_years) > 0:
+        filters["reporting_fiscal_year__in"] = fiscal_years
+    if federal_account_id:
+        pass
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/submissions/helpers.py]: Removed

File: usaspending_api/accounts/views/tas.py [hunk 4]
```diff
@@ -35,7 +36,12 @@ class TASBalancesQuarterAggregate(FilterQuerysetMixin, AggregateQuerysetMixin, C
     serializer_class = AggregateSerializer
 
     def get_queryset(self):
-        queryset = AppropriationAccountBalances.objects.all()
+        fiscal_years = self.get_fiscal_years()
+        federal_account_id = self.get_federal_account_id()
+
+        submission_ids = get_latest_submission_ids_for_each_fiscal_quarter(fiscal_years, federal_account_id)
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/accounts/views/tas.py]: Extra methods have been added to choose between file a/b

File: usaspending_api/submissions/helpers.py [hunk 5]
```diff
@@ -198,3 +198,40 @@ def get_latest_submission_ids_for_fiscal_year(fiscal_year: int):
         .values_list("submission_id", flat=True)
     )
     return submission_ids
+
+
+def get_latest_submission_ids_for_each_fiscal_quarter(
+    is_file_a, fiscal_years: List[int] = [], federal_account_id: int = None
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/submissions/helpers.py]: Extra methods have been added to make this more clear

File: usaspending_api/common/mixins.py [hunk 6]
```diff
@@ -212,6 +212,30 @@ def order_records(self, request, *args, **kwargs):
         else:
             return queryset
 
+    def get_federal_account_id(self):
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/common/mixins.py]: These have been combined into a single method

File: usaspending_api/accounts/views/tas.py [hunk 7]
```diff
@@ -35,7 +36,12 @@ class TASBalancesQuarterAggregate(FilterQuerysetMixin, AggregateQuerysetMixin, C
...
 
     def get_queryset(self):
-        queryset = AppropriationAccountBalances.objects.all()
+        fiscal_years = self.get_fiscal_years()
+        federal_account_id = self.get_federal_account_id()
+
+        submission_ids = get_latest_submission_ids_for_each_fiscal_quarter(True, fiscal_years, federal_account_id)
+
+        queryset = AppropriationAccountBalances.objects.filter(submission_id__in=submission_ids)
         queryset = self.filter_records(self.request, queryset=queryset)
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/accounts/views/tas.py:47]: I figured out why the `IN` was being added twice. I updated the queryset, so it is only included once. I also updated some pagination logic, so those endpoints don't include the `SELECT (1)` query

[the remark this answers is not in the document]
Comment (sethstoudenmier) [reply usaspending_api/accounts/views/tas.py:47]: @collinwr the queries look so much better and performance is a lot better even with Django_Debug running! 🎉

File: usaspending_api/submissions/helpers.py [hunk 8]
```diff
@@ -198,3 +198,55 @@ def get_latest_submission_ids_for_fiscal_year(fiscal_year: int):
...
+def _get_latest_submission_ids_for_each_fiscal_quarter(
+    federal_account_id_filter_obj, fiscal_years: List[int], federal_account_id: int
+):
+    filters = {"submission_window__submission_reveal_date__lte": now()}
+    if len(fiscal_years) > 0:
+        filters["reporting_fiscal_year__in"] = fiscal_years
+    if federal_account_id:
+        filters[f"{federal_account_id_filter_obj}__treasury_account__federal_account_id"] = federal_account_id
+
+    cte = With(
```
Comment (sethstoudenmier) [inline usaspending_api/submissions/helpers.py:212]: How does this work if you are selecting multiple Fiscal Years? The `latest_fiscal_period` is different for 2021 and 2020.

Comment (collinwr) [reply usaspending_api/submissions/helpers.py:212]: This results in a `GROUP BY` on `toptier_code`, `reporting_fiscal_year`, and `reporting_fiscal_quarter`, so it's getting the max `reporting_fiscal_period` for each Quarter of each Fiscal Year

**Review by sethstoudenmier — APPROVED** [review]

**Pull request merged**